### ISSUE TITLE
BEGIN/END probes on systems without uprobes

### ISSUE THREAD

**Issue by joelagnel** [iovisor/bpftrace#360]
bpftrace works great, however when I run certain scripts, it looks for uprobe_events in debugfs and fails.
```
# bpftrace /usr/share/bpftrace/tools/vfsstat.bt 
Attaching 8 probes...
open(/sys/kernel/debug/tracing/uprobe_events): No such file or directory
Error attaching probe: END
```
Could we make uprobe_events optional if a script doesn't need it? `vfsstat.bt` appears to only need kprobe support. Why does it try to open the uprobe_events file?  CC @brendangregg

**Comment by brendangregg** [comment]
vfsstat.bt uses a BEGIN probe, which I believe is implemented using uprobes: the BEGIN_trigger() function in triggers.h.

The BEGIN and END probes need a way to run some BPF code in kernel context, hence instrumenting its own user-space function that was created for that purpose. If we're to support systems that don't have uprobes, we'll need some kprobe/tracepoint we can trigger to do the same. I'm not sure there's a clean way to do that, so it'd be a matter of how dirty it is.

As a bad example, we could instrument syscalls:sys_enter_open, and if it's from bpftrace and is opening a file called /dev/bpftraceprobes/BEGIN (which will fail since it doesn't exist), then do the BEGIN routine. This would be terrible as we're adding overhead to all open()s, and doing a string comparison.

As a less-bad example, we could kprobe perf_event_open(), since we can call it from user space, it should be infrequent, and then pass in invalid arguments that we detect. Still not nice.

How common would it be to not have uprobes? If common, then maybe the way forward is for bpftrace to use the nice uprobe BEGIN_trigger function first if uprobes exist, and then use a hacky kprobe as a backup.

**Comment by joelagnel** [comment]
Thanks, I am curious why the BEGIN probe is even needed though. Most (all?) BEGIN probes seem to be just printing something. Can the implementation not print the header output directly, like BCC does? BCC doesn't need such probes.

Enable uprobes should be possible, it is not that common at the moment on Android because no one is using it. But kprobes is common. Thanks.

**Comment by brendangregg** [comment]
The BEGIN probe can do BPF things like initialize BPF map variables, eg, a "@start = nsecs" so that we can do walltimestamp calculations later in the script.

**Comment by joelagnel** [comment]
Got it, thanks

**Comment by mmarchini** [comment]
Ideally, for BEGIN and END we should be able to just run a BPF program without using probes or tracepoints. This might require some kernel work, as I don't think there's a way to do that today.

As an alternative, maybe we could use https://github.com/iovisor/ubpf for BEGIN/END?

**Comment by brendangregg** [comment]
Right, we need a bpf syscall call to say: run this now.

AFAIK, ubpf will be in its own memory space, and can't coordinate with the kernel one.

**Comment by vt-alt** [comment]
In [INSTALL.md](https://github.com/iovisor/bpftrace/blob/ac197439ee3fc8929d9c56897ae35685848e9813/INSTALL.md) you don't say that kernel should have `CONFIG_UPROBE_EVENTS` enabled for `bpftrace` to work.

**Comment by vt-alt** [comment]
Also, on ALT Linux since kernel 4.11 up to the current, `CONFIG_UPROBE_EVENTS` is not enabled. While this will be possibly changed in the future, `bpftrace` is not usable on older installations.

**Comment by ajor** [comment]
Fixed by #2264
